possibility to change name in lobby

Laserlicht opened the pull request:
fixes #2196

Nordsoft91 commented on server/CVCMIServer.cpp:
```diff
@@ -835,6 +835,19 @@ void CVCMIServer::setPlayer(PlayerColor clickedColor)
 
+void CVCMIServer::setPlayerName(PlayerColor color, std::string name)
+{
+	PlayerSettings & player = si->playerInfos[color];
```
```suggestion
	PlayerSettings & player = si->playerInfos.at(color);
```

Also, maybe test if color is valid player (CANNOT_DETERMINE is default value in cpack)

Nordsoft91 commented on server/CVCMIServer.cpp:
```diff
@@ -835,6 +835,19 @@ void CVCMIServer::setPlayer(PlayerColor clickedColor)
+
+	if(player.isControlledByHuman())
+	{
+		int nameID = *(player.connectedPlayerIDs.begin()); //if not AI - set appropiate ID
```
what if player disconnected from lobby? make sense to check for empty?

Nordsoft91 commented:
3 bugs during testing:
 - hotseat: start changing name of player1 -> click on name field player2 -> change name of player2 -> press enter
 result: only player2 name is updated
 - (more generic) player name is updated only if return key pressed. Change name without pressing return key -> start game
 result: name is not updated
 hint: maybe update player name on event of text field change (see chess timers fields as an example)
- multiplayer: only host player can change the name of themselves. Name for joined players cannot be changed neither by host nor by client player

Laserlicht commented:
>     * hotseat: start changing name of player1 -> click on name field player2 -> change name of player2 -> press enter
>       result: only player2 name is updated

fixed

>     * (more generic) player name is updated only if return key pressed. Change name without pressing return key -> start game
>       result: name is not updated
>       hint: maybe update player name on event of text field change (see chess timers fields as an example)

Yes. Tested this first. But it was not good when every keypress is send to server. Another problem is that after name change the lobby is updated (and the input field is also updated and lose the focus). I added a change if player clicks everywhere (other than the input field).
 
>     * multiplayer: only host player can change the name of themselves. Name for joined players cannot be changed neither by host nor by client player

Yes I disabled it (see `CSH->isGuest()`). Because the UI is locked (It's not possible to do enter and also not possible to send name change request to server). And honestly in Multiplayer it does make really sense because you asked for your name directly before the lobby. In Singleplayer you can currently change the name only with workarounds.

But if anyone find a solution for this in Multiplayer it's no problem to change it.

Nordsoft91 commented:
> And honestly in Multiplayer it does make really sense


Then maybe two options would be better than current behavior:
1. Allow host to change names for all guests
2. Disable name changing for host in MP games

Laserlicht commented:
used second approach

Nordsoft91 commented:
some wierd bug:
- start hotseat with 2 players
- rename player1
- click on player2 name
- start typing
- observe hotkeys pressed and focus lost

Laserlicht commented:
Yes. I know this. But it's a minor thing I think. I was not able to solve it.

Problem here is:
- rename player 1
- with clicking on player 2 name change on player 1 is accepted and send to server
- rename of player 2 is activated
- (few milliseconds later) server send new player names back -> this deactivate all text boxes
- text input isn't on textbox anymore -> hotkeys working again

Nordsoft91 reviewed (dismissed):
Change breaks chess-timers setup functionality:
- enable vcmi-extras
- in options tab, press timer selection button
- select chess timer
- try to change timer limits

Laserlicht commented:
@Nordsoft91 fixed; thanks!

Nordsoft91 commented on client/CServerHandler.h:
```diff
@@ -67,6 +67,7 @@ class IServerAPI
 	virtual void setCampaignBonus(int bonusId) const = 0;
 	virtual void setMapInfo(std::shared_ptr<CMapInfo> to, std::shared_ptr<CMapGenOptions> mapGenOpts = {}) const = 0;
 	virtual void setPlayer(PlayerColor color) const = 0;
+	virtual void setPlayerName(PlayerColor color, std::string name) const = 0;
```
```suggestion
	virtual void setPlayerName(PlayerColor color, const std::string & name) const = 0;
```

Nordsoft91 commented on client/CServerHandler.h:
```diff
@@ -152,6 +153,7 @@ class CServerHandler : public IServerAPI, public LobbyInfo
 	void setCampaignBonus(int bonusId) const override;
 	void setMapInfo(std::shared_ptr<CMapInfo> to, std::shared_ptr<CMapGenOptions> mapGenOpts = {}) const override;
 	void setPlayer(PlayerColor color) const override;
+	void setPlayerName(PlayerColor color, std::string name) const override;
```
```suggestion
	void setPlayerName(PlayerColor color, const std::string & name) const override;
```

Nordsoft91 commented on client/lobby/OptionsTab.h:
```diff
@@ -186,8 +188,14 @@ class OptionsTab : public InterfaceObjectConfigurable
 	private:
 		const OptionsTab & parentTab;
+
+		void changeName();
```
```suggestion
		void updateName();
```

Nordsoft91 commented on client/lobby/OptionsTab.cpp:
```diff
@@ -1115,6 +1123,43 @@ OptionsTab::PlayerOptionsEntry::PlayerOptionsEntry(const PlayerSettings & S, con
+		changeName();
+}
+
+void OptionsTab::PlayerOptionsEntry::changeName() {
```
```suggestion
void OptionsTab::PlayerOptionsEntry::updateName() {
```

Nordsoft91 commented on server/CVCMIServer.cpp:
```diff
@@ -835,6 +835,25 @@ void CVCMIServer::setPlayer(PlayerColor clickedColor)
+	if(!player.isControlledByHuman())
+		return;
+
+	if(!player.connectedPlayerIDs.size())
```
```suggestion
	if(player.connectedPlayerIDs.empty())
```
